src/xen_bus.c

```diff
@@ -0,0 +1,771 @@
+		if (domain->domid == iter->domid) {
+			if (fire_watcher(domain, id, iter->key)) {
+				k_free(iter->key);
+				sys_dlist_remove(&iter->node);
+				k_free(iter);
+				k_mutex_unlock(&pfl_mutex);
+			}
+		}
+	}
+
+	k_mutex_unlock(&pfl_mutex);
+}
+
+void handle_control(struct xen_domain *domain, uint32_t id, char *payload, uint32_t len)
+{
+	send_reply(domain, id, XS_CONTROL, "OK");
+}
+
+void handle_get_perms(struct xen_domain *domain, uint32_t id, char *payload, uint32_t len)
+{
+	send_errno(domain, id, ENOSYS);
+}
+
+void handle_set_perms(struct xen_domain *domain, uint32_t id, char *payload, uint32_t len)
+{
+	send_reply(domain, id, XS_SET_PERMS, "OK");
+}
+
+void remove_watch_entry(struct watch_entry *entry)
+{
+	k_free(entry->key);
+	k_free(entry->token);
+	sys_dlist_remove(&entry->node);
+	k_free(entry);
+}
+
+void handle_reset_watches(struct xen_domain *domain, uint32_t id, char *payload, uint32_t len)
+{
+	struct watch_entry *iter, *next;
+
+	k_mutex_lock(&wel_mutex, K_FOREVER);
+	SYS_DLIST_FOR_EACH_CONTAINER_SAFE (&watch_entry_list, iter, next, node) {
+		remove_watch_entry(iter);
+	}
+	k_mutex_unlock(&wel_mutex);
+
+	send_reply(domain, id, XS_RESET_WATCHES, "OK");
+}
+
+void handle_read(struct xen_domain *domain, uint32_t id, char *payload, uint32_t len)
+{
+	const char localpath[] = "/";
+	char path[128];
+
+	if (memcmp(payload, localpath, strlen(localpath)) == 0) {
+		memcpy(path, payload, strlen(payload) + 1);
+	} else {
+		snprintf(path, 128, "/local/domain/%d/%s", domain->domid, payload);
+	}
+
+	struct xs_entry *entry = key_to_entry(path);
+
+	if (entry) {
+		send_reply_read(domain, id, XS_READ, entry->value ? entry->value : "");
+		return;
+	}
+
+	send_reply(domain, id, XS_ERROR, "ENOENT");
+}
+
+void remove_recurse(sys_dlist_t *chlds)
+{
+	struct xs_entry *entry, *next;
+	SYS_DLIST_FOR_EACH_CONTAINER_SAFE (chlds, entry, next, node) {
+		if (entry->key) {
+			k_free(entry->key);
+			entry->key = NULL;
+		}
+
+		if (entry->value) {
+			k_free(entry->value);
+			entry->value = NULL;
+		}
+
+		remove_recurse(&entry->childs);
+
+		sys_dlist_remove(&entry->node);
+		k_free(entry);
+	}
+}
+
+void handle_rm(struct xen_domain *domain, uint32_t id, char *payload, uint32_t len)
+{
+	struct xs_entry *entry = key_to_entry(payload);
+
+	if (entry) {
+		if (entry->key) {
+			k_free(entry->key);
+			entry->key = NULL;
+		}
+
+		if (entry->value) {
+			k_free(entry->value);
+			entry->value = NULL;
+		}
+
+		k_mutex_lock(&xsel_mutex, K_FOREVER);
+		sys_dlist_remove(&entry->node);
+		sys_dlist_t chlds = entry->childs;
+		k_free(entry);
+		k_mutex_unlock(&xsel_mutex);
+
+		remove_recurse(&chlds);
+	}
+
+	send_reply_read(domain, id, XS_RM, "");
+}
+
+void handle_watch(struct xen_domain *domain, uint32_t id, char *payload, uint32_t len)
+{
+	const char localpath[] = "/";
+	char path[128];
+	char token[128];
+	uint32_t plen = 0;
+	bool is_relative = memcmp(payload, localpath, strlen(localpath)) != 0;
```

Comment:
Why not `payload[0] != '/'`?

Also, why some functions do check for a relative path, and other - do not?